src/vinylcontrol/vinylcontrolxwax.cpp

```diff
@@ -793,22 +795,72 @@ bool VinylControlXwax::uiUpdateTime(double now) {
     return false;
 }
 
-void VinylControlXwax::establishQuality(bool quality_sample) {
-    m_bQualityRing[m_iQualPos] = quality_sample;
-    if (m_iQualFilled < QUALITY_RING_SIZE) {
-        m_iQualFilled++;
+int VinylControlXwax::getPositionQuality() {
+    int positionQualityPercent;
+
+    if (m_iPosition == -1) {
+        // No position code - happens when control vinyl is spinning slow (or during scratching) or the setup is incorrect
+        positionQualityPercent = 0;
+    } else if ((m_iPosition <= m_iQualityLastPosition - 5) ||
+            (m_iPosition >= m_iQualityLastPosition + 5)) {
+        // Position code changed not by more than 5. This indicates a normal spinning conrol vinyl.
+        positionQualityPercent = 100;
+    } else {
+        // Position code changed by more than 5. This indicates a fast spinning control vinyl, a jumping needle or a false signal.
+        // This covers also the case of m_iQualityLastPosition == -1
+        positionQualityPercent = 50;
+    }
+
+    m_iQualityLastPosition = m_iPosition;
+
+    return positionQualityPercent;
+}
+
+int VinylControlXwax::getPitchQuality(double& pitch) {
+    int pitchQualityPercent;
+
+    if (m_iQualityRingFilled == 0) {
+        m_dQualityLastPitch = pitch;
+        pitchQualityPercent = 0;
+        return pitchQualityPercent; // First value - but at least two values needed for calculation
     }
 
-    int quality = 0;
-    for (int i = 0; i < m_iQualFilled; ++i) {
-        if (m_bQualityRing[i]) {
-            quality++;
+    double pitchDifference = pitch - m_dQualityLastPitch;
+    m_dQualityLastPitch = pitch;
+
+    if (pitchDifference != 0) {
+        double pitchStability = std::fabs(pitch / pitchDifference);
+
+        if (pitchStability < 3.0) {
+            // Unlikely that this pitch difference is from a proper set up control vinyl
+            pitchQualityPercent = 0;
+        } else if (pitchStability > 6.0) {
+            // Typical pitch difference for normal spinning control vinyl
+            pitchQualityPercent = 100;
+        } else {
+            // Weak pitch difference for a control vinyl slow spinning ,or for scratching
+            pitchQualityPercent = 75;
         }
+    } else {
+        // Unlikely case of two identical float values for pitch
+        pitchQualityPercent = 0;
+    }
+    return pitchQualityPercent;
+}
+
+void VinylControlXwax::establishQuality(double& pitch) {
+    m_iQualityRing[m_iQualityRingIndex] = getPositionQuality() + getPitchQuality(pitch);
+    m_fTimecodeQuality =
+            static_cast<float>(std::accumulate(
+                    m_iQualityRing, m_iQualityRing + m_iQualityRingFilled, 0)) /
+            static_cast<float>(2.0) / static_cast<float>(100.0) /
```

Comment:
is it necessary to static_cast literals to float?

Reply:
or, do the whole calculation as double and then static_cast the final result to float to reduce error

Reply:
You can write the literal as `2.0f` instead of `2.0` to make it float, not double.

Reply:
Changed this! Thanks forthe hint, I wasn't aware of this simple notation.